// @ng-bootstrap/ng-bootstrap/esm2020/offcanvas/offcanvas.module.mjs
import { NgModule } from '@angular/core';
import * as i0 from "@angular/core";
export { NgbOffcanvas } from './offcanvas';
export { NgbOffcanvasConfig } from './offcanvas-config';
export { NgbOffcanvasRef, NgbActiveOffcanvas } from './offcanvas-ref';
export { OffcanvasDismissReasons } from './offcanvas-dismiss-reasons';
export class NgbOffcanvasModule {
}
NgbOffcanvasModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: NgbOffcanvasModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgbOffcanvasModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.0", ngImport: i0, type: NgbOffcanvasModule });
NgbOffcanvasModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: NgbOffcanvasModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0", ngImport: i0, type: NgbOffcanvasModule, decorators: [{
            type: NgModule,
            args: [{}]
        }] });
