// @ngneat/transloco/esm2020/lib/transloco.providers.mjs
import { TRANSLOCO_LOADER } from './transloco.loader';
import { TRANSLOCO_CONFIG } from './transloco.config';
import { TRANSLOCO_SCOPE } from './transloco-scope';
import { DefaultTranspiler, TRANSLOCO_TRANSPILER, } from './transloco.transpiler';
import { DefaultHandler, TRANSLOCO_MISSING_HANDLER, } from './transloco-missing-handler';
import { DefaultInterceptor, TRANSLOCO_INTERCEPTOR, } from './transloco.interceptor';
import { DefaultFallbackStrategy, TRANSLOCO_FALLBACK_STRATEGY, } from './transloco-fallback-strategy';
export const defaultProviders = [
    {
        provide: TRANSLOCO_TRANSPILER,
        useClass: DefaultTranspiler,
        deps: [TRANSLOCO_CONFIG],
    },
    {
        provide: TRANSLOCO_MISSING_HANDLER,
        useClass: DefaultHandler,
    },
    {
        provide: TRANSLOCO_INTERCEPTOR,
        useClass: DefaultInterceptor,
    },
    {
        provide: TRANSLOCO_FALLBACK_STRATEGY,
        useClass: DefaultFallbackStrategy,
        deps: [TRANSLOCO_CONFIG],
    },
];
export function provideTransloco(options) {
    const providers = [...defaultProviders];
    if (options.config) {
        providers.push(...provideTranslocoConfig(options.config));
    }
    if (options.loader) {
        providers.push(...provideTranslocoLoader(options.loader));
    }
    return providers;
}
export function provideTranslocoConfig(config) {
    return [
        {
            provide: TRANSLOCO_CONFIG,
            useValue: config,
        },
    ];
}
export function provideTranslocoLoader(loader) {
    return [{ provide: TRANSLOCO_LOADER, useClass: loader }];
}
export function provideTranslocoScope(scope) {
    return [
        {
            provide: TRANSLOCO_SCOPE,
            useValue: scope,
        },
    ];
}
